File: Python3/135. Candy.py
class Solution:
    def count(self, n):
        return (n*(n+1)) // 2;
    
    def candy(self, ratings: List[int]) -> int:
        candies = old = up = down = 0
        
        for i, r in enumerate(ratings[1:], 1):
            if ratings[i-1] > r:
                new = -1
            elif ratings[i-1] < r:
                new = 1
            else:
                new = 0
            # reach a peak or a valley    
            if (old > 0 and new == 0) or (old < 0 and new >= 0):
                candies += self.count(up) + self.count(down) + max(up, down)
                up = down = 0
            
            if new > 0:
                up += 1
            elif new < 0:
                down += 1
            else:
                candies += 1
            old = new
            
        candies += self.count(up) + self.count(down) + max(up, down) + 1
        return candies

File: Java/135. Candy.java
class Solution {
    private int count (int n) {
        return (n * (n+1)) / 2;
    }
    public int candy(int[] ratings) {
        int up = 0, down = 0, olds = 0, candies = 0, news;
        for (int i=1; i<ratings.length; i++) {
            news = (ratings[i-1] < ratings[i]) ? 1 : (ratings[i-1] > ratings[i]) ? -1 : 0;
            
            if ((olds > 0 && news == 0) || (olds < 0 && news >= 0)) {
                candies += count(up) + count(down) + Math.max(up, down);
                up = 0;
                down = 0;
            }
            
            if (news > 0)   up++;
            else if (news < 0)  down++;
            else   candies++;
            
            olds = news;
        }
        
        candies += count(up) + count(down) + Math.max(up, down) + 1;
        return candies;
    }
}